contracts/RewardController.sol

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: MIT
+// Disclaimer https://github.com/hats-finance/hats-contracts/blob/main/DISCLAIMER.md
+
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "./interfaces/IHATVault.sol";
+import "./interfaces/IRewardController.sol";
+
+error EpochLengthZero();
+
+contract RewardController is IRewardController, OwnableUpgradeable {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    // Not enough rewards to transfer to user
+    error NotEnoughRewardsToTransferToUser();
+
+    struct VaultInfo {
+        uint256 rewardPerShare;
+        uint256 lastProcessedVaultUpdate;
+        uint256 lastRewardBlock;
+        uint256 allocPoint;
+    }
+
+    struct VaultUpdate {
+        uint256 blockNumber;// update blocknumber
+        uint256 totalAllocPoint; //totalAllocPoint
+    }
+
+    uint256 public constant REWARD_PRECISION = 1e12;
+    uint256 public constant NUMBER_OF_EPOCHS = 24;
+
+    // Block from which the vaults contract will start rewarding.
+    uint256 public startBlock;
+    uint256 public epochLength;
+    // the ERC20 contract in which rewards are distributed
+    IERC20Upgradeable public rewardToken;
```

Comment:
why does it have to be upgradeable?

Reply:
It  does not really matter, but perhaps using IERC20 is slightly more elegant

Reply:
i think so

Reply:
actually i tried it and got `Error: Contract `RewardController` is not upgrade safe` .. so nevermind :)